File: leetcode/algorithms/404_sum_of_left_leaves/main.java
import java.util.Stack;

class SumOfLeftLeaves {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int sumOfLeftLeaves(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int result = 0;

        if (root.left != null) {
            if (root.left.left == null && root.left.right == null) {
                result += root.left.val;
            } else {
                result += sumOfLeftLeaves(root.left);
            }
        }

        result += sumOfLeftLeaves(root.right);

        return result;
    }


    // Solution
    // Solution 1: Recursive
    public int solution1(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int ans = 0;
        if (root.left != null) {
            if (root.left.left == null && root.left.right == null) {
                ans += root.left.val;
            } else {
                ans += solution1(root.left);
            }
        }

        ans += solution1(root.right);

        return ans;
    }

    // Solution 2: Iterative
    public int solution2(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int ans = 0;

        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);

        while (!stack.empty()) {
            TreeNode node = stack.pop();

            if (node.left != null) {
                if (node.left.left == null && node.left.right == null) {
                    ans += node.left.val;
                } else {
                    stack.push(node.left);
                }
            }

            if (node.right != null) {
                if (node.right.left != null || node.right.right != null) {
                    stack.push(node.right);
                }
            }
        }

        return ans;
    }
}

File: leetcode/algorithms/404_sum_of_left_leaves/main.c
struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

// My Solution 1
int sumOfLeftLeaves1(struct TreeNode* root) {
    int sum = 0;

    if (root) {
        if (root->left) {
            if (!root->left->left && !root->left->right) {
                sum += root->left->val;
            }
        }
        sum += sumOfLeftLeaves1(root->left) + sumOfLeftLeaves1(root->right);
    }

    return sum;
}

// My Solution 2: with flag
int addLeftVal(struct TreeNode* root, int isLeft) {
    int sum = 0;

    if (root) {
        if (!root->left && !root->right && isLeft) {
            sum += root->val;
        } else {
            sum += addLeftVal(root->left, 1) + addLeftVal(root->right, 0);
        }
    }

    return sum;
}

int sumOfLeftLeaves2(struct TreeNode* root) {
    return addLeftVal(root->left, 1) + addLeftVal(root->right, 0);
}
